File: haskell/src/Solutions/Euler078.hs
module Solutions.Euler078 where
import Utils.Array(funArray)
import Data.Array

-- Pentagonal number theorem
pentagonal :: [(Int, Int)]
pentagonal = pen 1 where
    pen k = [((-1)^(k - 1), k * (3 * k - 1) `div` 2),
             ((-1)^(k - 1), k * (3 * k + 1) `div` 2)] ++ pen (k + 1)

bound = 75000 :: Int
partitions :: Array Int Int
partitions = funArray 0 bound f

f :: Int -> Int
f 0 = 1
f n = sum (map (\(mul, y) -> mul * partitions ! (n - y))
    $ takeWhile ((<=n) . snd) pentagonal) `mod` 1000000

euler078 :: IO String
euler078 = return $ show $ fst $ head $ filter ((==0) . snd) $ assocs partitions

File: rust/src/solutions/euler078.rs
pub struct Pentagonal {
    k: i64,
    sign: i64,
    first: bool,
}

impl Pentagonal {
    pub fn new() -> Pentagonal {
        Pentagonal {
            k: 1,
            sign: 1,
            first: true,
        }
    }
}

impl Iterator for Pentagonal {
    type Item = (i64, i64);
    fn next(&mut self) -> Option<Self::Item> {
        if self.first {
            self.first = false;
            Some((self.sign, self.k * (3 * self.k - 1) / 2))
        } else {
            let ans = (self.sign, self.k * (3 * self.k + 1) / 2);
            self.k += 1;
            self.sign = -self.sign;
            self.first = true;
            Some(ans)
        }
    }
}

pub fn euler078() -> String {
    let bound: usize = 75000;
    let mut partitions = vec![1i64; bound];
    for n in 1..bound {
        partitions[n] = Pentagonal::new()
            .take_while(|&(_, k)| k <= n as i64)
            .map(|(sign, k)| sign * partitions[n - k as usize])
            .sum::<i64>()
            % 1000000;
    }
    (1usize..)
        .find(|&i| partitions[i] == 0)
        .unwrap()
        .to_string()
}
